tasty-query/shared/src/main/scala/tastyquery/reader/tasties/TreeUnpickler.scala

```diff
@@ -1064,6 +1065,24 @@ private[tasties] class TreeUnpickler private (
     caches.getSymbol[Symbol](symAddr)
   }
 
+  /** Reads a package reference, with a fallback on faked term references.
+    *
+    * In a full, correct classpath, `readPackageRef()` will always return a
+    * `PackageRef`. However, in an incomplete or incorrect classpath, this
+    * method may return a `TermRef` if the target package does not exist.
+    *
+    * An alternative would be to create missing packages on the fly, but that
+    * would not be consistent with `Trees.Select.tpe` and
+    * `Trees.TermRefTypeTree.toType`.
+    */
+  private def readPackageRef(): PackageRef | TermRef =
```

Comment:
or create the package here?

Reply:
If we create package symbols on-demand, even if marked as synthetic, we start having big problems if a synthetic package symbol clashes with an actual, non-package term symbol. Creating all package symbols first, then none later, is the only way to prevent non-determinism in that space, AFAICT. That means not having symbols for should-be-`PackageRef`s that don't actually exist.



> this is arguably pretty bad because syntactically its a package ref, and we wanted to make syntax non-context dependent. what about making PackageRef have an optional symbol at construction? (so still accessible without context)

Sort of, but we have precedent as far as package refs are concerned. When unpickling Scala 2 pickles, for example, the corresponding encoding already behaves in what this PR brings to .tasty files. Indeed, in Scala 2, there is only "external symbol ref" with a path that starts from root. We eagerly follow packages that exist to create `PackageRef`s, and once we don't find a package anymore, we proceed with `TermRef`s. This is not really different.

In practice, error messages will be decent, as we will anyway end up with `MemberNotFoundException`s when we try to access the dangling `TermRef`s'symbols.

Reply:
ok, that is reasonable, (user should supply the correct classpath anyway)